web/ww.ts

```diff
@@ -135,215 +270,106 @@ class Wormhole {
+		const stats = await this.pc.getStats();
+
+		let id: string | undefined;
+		stats.forEach((s) => {
+			// s.selected gives more confidenece than s.state == "succeeded", but Chrome does
+			// not implement it.
+			if (
+				s.type === "candidate-pair" &&
+				(s as RTCIceCandidatePairStats).state === "succeeded"
+			) {
+				id = (s as RTCIceCandidatePairStats).localCandidateId;
 			}
+		});
 
-			case "wait_for_webtc_answer": {
-				if (!this.key || !this.pc) { return }
+		if (!id) {
+			return "";
+		}
 
-				const msg: RTCSessionDescriptionInit | null = JSON.parse(webwormhole.open(this.key, m.data));
-				if (!msg) {
-					this.fail("bad key");
-					this.ws.send(webwormhole.seal(this.key, "bye"));
-					this.ws.close(WormholeErrorCodes.closeBadKey);
-					return;
-				}
-				if (msg.type !== "answer") {
-					console.log("unexpected message", msg);
-					this.fail("unexpected message");
-					return;
-				}
-				console.log("got answer");
-				this.pc.setRemoteDescription(new RTCSessionDescription(msg));
-				this.resolve3(webwormhole.fingerprint(this.key));
-				this.state = "wait_for_candidates";
-				return;
+		let conntype: string = "";
+		stats.forEach((s) => {
+			if (s.id === id) {
+				conntype = (s as { candidateType: string }).candidateType;
 			}
+		});
 
-			case "wait_for_candidates": {
-				if (!this.key || !this.pc) { return }
+		return conntype;
+	}
 
-				const msg: {candidate: string} | null = JSON.parse(webwormhole.open(this.key, m.data));
-				if (!msg) {
-					this.fail("bad key");
-					this.ws.send(webwormhole.seal(this.key, "bye"));
-					this.ws.close(WormholeErrorCodes.closeBadKey);
-					return;
-				}
-				console.log("got remote candidate", msg.candidate);
-				this.promise2.then(async () => {
-					if (!this.key || !this.pc) { return }
-					this.pc.addIceCandidate(new RTCIceCandidate(msg));
-				});
-				return;
-			}
+	async dial(): Promise<Uint8Array> {
+		this.ws = new WebSocket(
+			Wormhole.wsserver(this.signalserver, this.slot),
+			Wormhole.protocol
+		);
+		// Use lambdas so that 'this' in the respective bodies refers to the Wormhole
+		// instance, and not the WebSocket one.
+		this.ws.onopen = () => this.onopen();
+		this.ws.onerror = (e: Event) => this.onerror(e);
+		this.ws.onclose = (e: CloseEvent) => this.onclose(e);
+		this.ws.onmessage = (e: MessageEvent) => this.onmessage(e);
 
-			case "wait_for_pc_initialize":
-			case "wait_for_local_offer":
-			case "wait_for_local_answer": {
-				console.log("unexpected message", m);
-				this.fail("unexpected message");
-				return;
-			}
-			case "error":
-				return;
-		}
+		return this.done;
 	}
 
-	makePeerConnection(iceServers: GoICEServers) {
-		let normalisedICEServers = [];
-		for (let i = 0; i < iceServers.length; i++) {
-			normalisedICEServers.push({
-				urls: iceServers[i].URLs,
-				username: iceServers[i].Username,
-				credential: iceServers[i].Credential,
-			});
+	async onmessage(m: MessageEvent) {
```

Comment:
Just FYI, if you use this syntax, the `this` is effectively prebound to this instance. Perhaps your way of doing things is more explicit though.

```suggestion
	onmessage = async (m: MessageEvent) => {
```